### PR TITLE
Update Readme and heavily refactor

### PR DESCRIPTION
**Pull request by ianfixes**
# TL;DR
- tried to make stuff more readable / understandable
- tried to improve logical grouping of various things
- put redundant code into functions
- used more features of the language where it was helpful

**Full disclosure, I only tested this in my head because I didn't have any hardware.**
### Wiring

Wiring variables now start with `PIN_` and are at the top

``` arduino
const int PIN_JOY_X       = 0;
const int PIN_JOY_Y       = 1;
const int PIN_EN_BTN      = 10;
const int PIN_VALVE_X_POS = 3;
const int PIN_VALVE_X_NEG = 5;
const int PIN_VALVE_Y_POS = 6;
const int PIN_VALVE_Y_NEG = 7;
```
### Named & Indexed Axes

Axes have now been named `axisX` and `axisY`.  All this means is that we defined some names to represent positions in an array.

``` arduino
const int axisX = 0;
const int axisY = 1;
```

So instead of 2 joystick variables, we use one array variable with 2 indexes.  

``` arduino
// Set up an array of joystick axes
ArduinoJoystick joystick[] = {
  ArduinoJoystick(0, PIN_JOY_X),
  ArduinoJoystick(1, PIN_JOY_Y)
};

joystick[axisX].setPoints(390, 520, 640);
joystick[axisY].setDeadbands(5, 5, 5);
```
### Setting valves

All valve-setting code has been put into its own function that takes an axis index and a desired value.  The proper behavior seems to be that only one valve pin (positive or negative) should be non-zero at any time.  When a zero is set, both pins should be zero.

``` arduino
void setValve(int axis, int valvePosition) {
  // Set the positive valve position to zero if the desired position is 0 (or less), otherwise the desired value
  analogWrite(valve[axis].pos, valvePosition <= 0 ? 0 : valvePosition);
  // Set the negative valve position to zero if the desired position is 0 (or more), otherwise the absolute desired value
  analogWrite(valve[axis].neg, valvePosition >= 0 ? 0 : abs(valvePosition));
}
```
### Main loop

The primary function of the code is to read values from the joystick and set the valves accordingly.  This is that:

``` arduino
  for (int axis = axisX; axis <= axisY; ++axis) {
   joystick[axis].poll();  // Read the latest value

    // constrain the joystick value, and only send it if the enable button is pressed (otherwise, 0)
    int stickPosition = constrain(joystick[axis].getValue(), -255, 255);
    setValve(axis, isEnabled ? stickPosition : 0);  
  }
```
# Language features

I took advantage of some features of the Arduino language that weren't in the original code, so (just in case they're unfamiliar) here is what they are and how they work:
### Trinary operator

Instead of this:

``` arduino
if (CONDITION) {
  x = 3;
} else {
  x = 7000;
}
```

Do this:

``` arduino
x = CONDITION ? 3 : 7000;
```
### Arrays and shorthand

Instead of this:

``` arduino
int voltage0 = 135;
int voltage1 = 148;
int voltage2 = 258;
```

You can use an array like this:

``` arduino
int voltage[3];  // a container with 3 slots
voltage[0] = 135;
voltage[1] = 148;
voltage[2] = 258;
```

And that can be expressed in shorthand (only works when you first declare the array):

``` arduino
int voltage[] = { 135, 148, 258 };
```
### Structs and shorthand

Sometimes several variables go together. 

``` arduino
int constraintAmin = 2;
int constraintAmax = 49;
int constraintBmin = 3;
int constraintBmax = 34;
```

Package them up in a struct.

``` arduino
struct Range {
  int min;
  int max;
}

Range constraintA;
Range constraintB;
constraintA.min = 2;
constraintA.max = 49;
constraintB.min = 3;
constraintB.max = 34;
```

_But wait, that's a lot more lines that add basically nothing!_  True.  There's a shorthand for assigning to structs, and it's the same as what happens for arrays.

``` arduino
struct Range {
  int min;
  int max;
}

Range constraintA = {2, 49};
Range constraintB = {3, 34};
```

### CODE REVIEW

**Comment by ifreecarve** [comment]
@DanNixon I think this fixes #7

File: HBR-Proportional-Valve/HBR-Proportional-Valve.ino [hunk 1]
```diff
@@ -2,151 +2,132 @@
...
-  {
-    pinMode(pwmPins[i], OUTPUT);
-    analogWrite(pwmPins[i], 0);
+  for (int axis = axisX; axis <= axisY; ++axis) {
+    joystick[axis].setPoints(390, 520, 640); //set Low/Center/High positions
+    joystick[axis].setDeadbands(5, 5, 5);    //set the deadbands
+    joystick[axis].setThreshold(5);          //set the minimum change to trigger an event
+ 
+    //set the transform from input values to output values
+    joystick[axis].setTransformation(joystickPotTransform);
```
Comment (DanNixon) [inline HBR-Proportional-Valve/HBR-Proportional-Valve.ino]: I would not do this, calibrations should be done per joystick rather than assuming both joysticks to be identical.

I have a pair of these joysticks and they do not share an identical calibration.

**Comment by DanNixon** [comment]
Looks good other than the comment I left.

Agree that it fixes #4.

**Comment by ifreecarve** [comment]
OK, thanks. I expanded out the joystick settings and added some more info on the coding style.

**Comment by Bumblefuck** [comment]
Very nice. I will try on a proper arduino tonight.
